File: patterns/creational_singleton/creational_singleton_webref.py

class Counter:
  def __init__(self, name):
    self.name = name
      
    self.count = 0

  def increase_count(self):
    self.count = self.count + 1
    # return


  def get_count(self):
    print(f"{self.name}: count: {self.count}")

if __name__ == "__main__":
  count1 = Counter('count1') 
  count2 = Counter('count2') 

  # пока мне не понятно
  # id-шки должны быть одинаковыми или разными 

  print(f"id count1: {id(count1)}")
  print(f"id count2: {id(count2)}")

  count1.increase_count()
  count1.increase_count()
  count2.increase_count()
  count1.increase_count()

  count1.get_count()
  count2.get_count()

File: patterns/creational_singleton/creational_singleton_webref.js
// #1

const instance1 = {
    name: "singleton"
}

const instance2 = {
    name: "singleton"
}

// console.log(instance1 === instance2); // false
// console.log({} === {}); // false

// #2
//// #bad

// let instance;

// class Counter {
//     constructor() {
//         if (!instance) instance = this;
//         instance.count = 0;
//         return instance;
//     }

//     getCount() {
//         return instance.count;
//     }

//     increaseCount() {
//         return instance.count++;
//     }
// }

//// #good
class Counter {
    constructor() {
        if (typeof Counter.instance === 'object')  {
            return Counter.instance;
        }

        // console.log(this);
        
        this.count = 0;
        Counter.instance = this;
        return this;
    }

    getCount() {
        return this.count;
    }

    increaseCount() {
        return this.count++;
    }
}

// ----
const myCount1 = new Counter();
const myCount2 = new Counter();

myCount1.increaseCount();
myCount1.increaseCount();
myCount2.increaseCount();
myCount2.increaseCount();

console.log(myCount2.getCount()); // 4
console.log(myCount1.getCount()); // 4

